dpctl/tensor/libtensor/include/kernels/sorting/topk.hpp

```diff
@@ -468,9 +475,16 @@ sycl::event topk_radix_impl(sycl::queue &exec_q,
     IndexTy *workspace = sycl::malloc_device<IndexTy>(
         padded_total_nelems + total_nelems, exec_q);
 
+    if (nullptr == workspace) {
+        throw std::runtime_error(
+            "Not enough device memory for radix sort topk");
+    }
+
     IndexTy *tmp_tp = sycl::malloc_device<IndexTy>(total_nelems, exec_q);
 
-    if (nullptr == workspace || nullptr == tmp_tp) {
+    if (nullptr == tmp_tp) {
+        using dpctl::tensor::alloc_utils::sycl_free_noexcept;
+        sycl_free_noexcept(workspace, exec_q);
         throw std::runtime_error(
             "Not enough device memory for radix sort topk");
     }
```

Comment:
In addition to this, the `parallel_radix_sort_impl` call also can raise `runtime_error` if it fails to allocate more additional memory.

It should be wrapped in a `try-except` block to account for that to avoid another memory leak here.

At some point it would be nice to use `unique_ptr` here and avoid this entirely, but that would be a more in-depth refactoring.

Reply:
The same change as above should be made in `radix_argsort_axis1_contig_impl` as well.

Reply:
This has been implemented. Lifetime of USM allocations in "sorting/" is handled by a `unique_ptr`, which is released after the `host_task` has been submitted.